### ISSUE TITLE
Incorrect call to java.lang.String#substring()

### ISSUE THREAD

**Issue by kristapsdz-saic** [jcoglan/canopy#60]
In the Java grammar, the generated code will try to cut a substring as in the following pseudocode:

```
if (offset < input.length)
    result := input.substring(offset, offset + xxx);
```

The substring may clearly be called if `offset + xxx` exceeds the input's length.  In Java, [substring](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#substring(int,%20int)) will throw an exception if this is the case, which will cause the code to abort if this condition is met.

The condition is met trivially by having a grammar with an expected string larger than the given one at the end of the file.

I've patched this in canopy.js as in the npm module, but am not sure how to do this for the source.

```
diff --git a/node_modules/canopy/lib/canopy.js b/node_modules/canopy/lib/canopy.js
index a575820..ae3c9a9 100644
--- a/node_modules/canopy/lib/canopy.js
+++ b/node_modules/canopy/lib/canopy.js
@@ -3186,7 +3186,7 @@ Canopy.extend(Canopy, {

     chunk_: function(length) {
       var chunk = this.localVar_('chunk', this.null_()), input = 'input', of = 'offset';
-      this.if_(of + ' < inputSize', function(builder) {
+      this.if_(of + ' + ' + length + ' <= inputSize', function(builder) {
         builder._line(chunk + ' = ' + input + '.substring(' + of + ', ' + of + ' + ' + length + ')');
       });
       return chunk;
```

In this patch, the full string size is tested before taking the conditional.  This fixes the generated code for me in several grammars.

**Comment by jcoglan** [comment]
This is a duplicate of #23 and is fixed in version 0.4.0.

**Comment by kristapsdz-saic** [comment]
Oops, I missed that!  Sorry for the churn, and thanks again for this great software!
